api/admin/controller/patron_auth_services.py

```diff
@@ -0,0 +1,172 @@
+                                SIP2AuthenticationProvider.__module__,
+                                FirstBookAuthenticationAPI.__module__,
+                               ]
+
+    def process_patron_auth_services(self):
+        self.require_system_admin()
+
+        if flask.request.method == 'GET':
+            return self.process_get()
+        else:
+            return self.process_post()
+
+    def process_get(self):
+        services = self._get_integration_info(ExternalIntegration.PATRON_AUTH_GOAL, self.protocols)
+        return dict(
+            patron_auth_services=services,
+            protocols=self.protocols,
+        )
+
+    def process_post(self):
+        protocol = flask.request.form.get("protocol")
+        is_new = False
+        error = self.validate_form_fields(protocol)
+        if error:
+            return error
+
+        id = flask.request.form.get("id")
+        if id:
+            # Find an existing service to edit
+            auth_service = get_one(self._db, ExternalIntegration, id=id, goal=ExternalIntegration.PATRON_AUTH_GOAL)
+            if not auth_service:
+                return MISSING_SERVICE
+            if protocol != auth_service.protocol:
+                return CANNOT_CHANGE_PROTOCOL
+        else:
+            # Create a new service
+            auth_service, is_new = self._create_integration(
+                self.protocols, protocol, ExternalIntegration.PATRON_AUTH_GOAL
+            )
+            if isinstance(auth_service, ProblemDetail):
+                return auth_service
+
+        name = self.get_name(auth_service)
+        if isinstance(name, ProblemDetail):
+            return name
+        elif name:
+            auth_service.name = name
+
+        [protocol] = [p for p in self.protocols if p.get("name") == protocol]
+        result = self._set_integration_settings_and_libraries(auth_service, protocol)
+        if isinstance(result, ProblemDetail):
+            return result
+
+        library_error = self.check_libraries(auth_service)
+        if library_error:
+            return library_error
+
+        if is_new:
+            return Response(unicode(auth_service.id), 201)
+        else:
+            return Response(unicode(auth_service.id), 200)
+
+    def validate_form_fields(self, protocol):
+        """Verify that the protocol which the user has selected is in the list
+        of recognized protocol options."""
+
+        if protocol and protocol not in [p.get("name") for p in self.protocols]:
+            return UNKNOWN_PROTOCOL
+
+    def get_name(self, auth_service):
+        """Check that there isn't already an auth service with this name"""
+
+        name = flask.request.form.get("name")
+        if name:
+            if auth_service.name != name:
+                service_with_name = get_one(self._db, ExternalIntegration, name=name)
+                if service_with_name:
+                    self._db.rollback()
```

Comment:
As with the other branch, the calls to rollback should be in the method that modified the database in the first place -- that is, before the return statements in process_post().